### ISSUE TITLE
NIO to allow GC during JNI and access to special memory regions

### ISSUE THREAD

**Issue by almson** [fommil/netlib-java#58]
It would be nice if netlib-java exposed NIO interfaces to its routines.

Direct ByteBuffers (and FloatBuffer, DoubleBuffer,  etc) have the following neat features:
- zero-copy between Java and native code,  giving 100% performance parity with C. 
- as fast as primitive arrays
- can memory-map files (Should be possible to do calculations directly on huge matrices residing on SSDs)
- compatibility with code that already uses them (I've been using jcuda and jcublas and use direct bytebuffers, and have to incure double copy overhead to use an array-based jni api)
- simple to expose

**Comment by dlwh** [comment]
(obviously this is low priority and you have probably less bandwidth than I do, but I am somewhat interested in support for NIO Buffers. It's come up a few times in issues and questions for Breeze.)

**Comment by fommil** [comment]
If you know anyone willing to fund this for a few months, please let me know.

**Comment by fommil** [comment]
BTW most people who think they need nio buffers don't really understand what netlib-java is doing and don't actually need it. But I can see it would be an improvement and there are rare cases that would benefit.

**Comment by bkgood** [comment]
I've got a fork over at https://github.com/bkgood/netlib-java that dirtily adds NIO buffer support. Would you be interested in merging it in if I polished it up?

It's got some extraneous stuff I want to remove, the F2J implementation fails on direct buffers and the JNI implementations fail on array-backed buffers. It currently fulfills my needs (multithreading done in the BLAS implementation wasn't cutting it so I'm doing in myself in-JVM; long story) but I'd love to get a cleaned up version in in a proper release if you'd be interested in such a version.

On fixing the latter two issues I brought up- I agree that letting the F2J implementation handle non-array-backed buffers will slow it down considerably, but given the incredibly poor performance this implementation currently exhibits, I'm not sure I see the extra copying work leading to much of a regression. If performance is a goal of the Java implementation, F2J seems to really miss the mark - I, at one point, reimplemented dgemv in Java in such a way that I had it running in the same order of magnitude as the OpenBLAS implementation because I was able to write Java in such a way that Hotspot was able to vectorize it. F2J's bytecode doesn't seem to be written (well, generated) in such a way. I am concerned about the necessary allocations, though. Any ideas here? Like I mentioned earlier, I currently just throw an exception which might be better than implicitly allocating a potentially huge array.

To fix the final issue (JNI implementation fails on array-backed buffers), it seems to make sense to just get the backing array from the buffer and call GetPrimitiveArrayCritical on it. To my knowledge there isn't a sort of nio buffer that has neither a 'peer' nor a backing array but this can also be guarded against. 

FWIW there are a couple of nice uses I see for supporting NIO buffers here but I've been unable to see gains in TTSP (time to safepoint, which ideally avoiding the pinning of GetPrimitiveArrayCritical would reduce, but I've yet to see evidence of any gains). The biggest gain has been in memory locality as all the hardware backing my current project is commodity multi-socked x64 boxes under NUMA constraints, achieved with the help of some topology-aware allocation routines.

**Comment by fommil** [comment]
@bkgood hi, thanks for looking at this! From skimming your branch, it does look a bit hacky for me to be able to merge into the main code branch: this would effectively create two extra API variants for everything that we currently have? (correct me if I'm wrong)

If this work is going to be done, I'd like it to be done from the ground up and for F2J to support the same API. That approach would involve going the #76 direction. Realistically, I'd be looking for funding for a 6 month period to do this.

**Comment by bkgood** [comment]
> From skimming your branch, it does look a bit hacky for me to be able to merge into the main code branch

Agreed, that's why I wanted to ask for your thoughts before doing a big code drop-style pull request ;)

> this would effectively create two extra API variants for everything that we currently have?

I was planning to delete the `POINTER_TO_LONG` variant as part of my cleanup (performance improvement over `NIO_BUFFER` was appreciable but not what I'd consider significant), so it would only be the  `NIO_BUFFER` variant plus the original array-based one.

> If this work is going to be done, I'd like it to be done from the ground up and for F2J to support the same API. That approach would involve going the #76 direction. Realistically, I'd be looking for funding for a 6 month period to do this.

Unfortunately my patches came from an immediate need and since they fulfill my needs (the F2J implementation is very much useless in my project, although I see its utility elsewhere) I can't really sell 6 months' funding and time.

One could presumably modify the F2J class files such that we duplicate all the methods (and associated bytecode), splice out the relevant `Xaload` instructions and splice in calls to `XBuffer.get` and replace the relevant bits of the method signatures. I'm less than excited about such a proposition for perhaps obvious reasons :) but I agree that fully supporting F2J is necessary for the goals of netlib-java. This would entail significantly less work IMO than a new Fortran-to-Java transpiler (and I would be willing to give it a go), although has a hacky nature of its own.

If you're interested in a patch that solves the issue in a way similar to what I've got now, I'd very much like to work to get something mergeable :) I'm more than willing to do the work to get it there.

**Comment by fommil** [comment]
I suppose I could be convinced by something that supports F2J in a second API. But actually cutting a belli is extremely complex. Perhaps it's easier now that docker exists, but OS X was always the problem.

**Comment by fommil** [comment]
/belli/release/ (to idea what autocorrect was doing there)

**Comment by bkgood** [comment]
A second API as in a second interface and set of implementations? Or additional methods to the existing class, along with support for copying in from DirectBuffers? (IOW an extension of what I'm doing now) My memory is a bit fuzzy but I imagine I could get this to generate a separate set of classes.

FWIW I've been developing on OSX and have my SNAPSHOT released building there so I could lend a hand as needed. I imagine you'd want to do the building since you're signing but I can at least share my experiences (homebrew in particular made it fairly easy to get the fortran compiler I needed, and then it was just a matter of fixing a few paths and commenting out the non-osx stuff from the various pom.xmls).

**Comment by fommil** [comment]
I think a second API would be best otherwise we'll be hitting classfile limits on the existing interfaces/classes.

**Comment by bkgood** [comment]
Sorry, your reply got buried in a gmail thread somehow :/

Looking at the classes generated by my fork, it seems we're quite far away from classfile limits on methods/class (https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.11 says ~65k)

    17:25 $ grep 'public native' native_system/java/target/netlib-native/com/github/fommil/netlib/NativeSystemARPACK.java |wc -l
          48
    17:28 $ grep 'public native' native_system/java/target/netlib-native/com/github/fommil/netlib/NativeSystemLAPACK.java |wc -l
        2352
    17:29 $ grep 'public native' native_system/java/target/netlib-native/com/github/fommil/netlib/NativeSystemBLAS.java |wc -l
         404

It seems ok from a classfile perspective (and these numbers actually include the methods taking a pointer in a long which I don't want to merge, so they're actually higher than they ought to be).

**Comment by fommil** [comment]
Well, there is that limit, but I'm also thinking about the burden on anyone who uses this library.

**Comment by eiennohito** [comment]
ByteBuffer JNI stubs can be similar to what lz4-java does for array/BB support (https://github.com/jpountz/lz4-java/blob/master/src/jni/net_jpountz_lz4_LZ4JNI.c#L94)

**Comment by almson** [comment]
@bkgood It's exciting to hear that you were about to get the JVM to vectorize your methods. I've been following the autovectorization developments, but in my experience only loops that index the array starting from 0 get vectorized. Were you able to get it working with non-0 offsets? Or is your dgemv limited in that sense?
